### PR TITLE
feat/palette-json-generator-webapp

### PR DESCRIPTION
**Pull request by klown**
A web-app for entering glosses, BCI AV IDs, and svg builder strings to produce and save a palette.

The main point is to make it easier than the previous way to make palettes by using  text editor to edit the JSON description file directly.  The previous way was time consuming and very error prone.  The web-app approach allows most of the generation of the JSON to be automated, and allows the user to later tweak the JSON file as necessary.

This branch was originally created based on the code for PR #21, but incorporated @cindyli's script into user interface.  It's not clear if this should be directly merged with the main branch or merged into the feat/palette-json-generator script branch first.

### CODE REVIEW

File: apps/palette-generator/palette-generator.html [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+      </p>
+      <p>
+        <label for="colStart">Starting column (>=1):</label><br>
+        <input type="number" id="colStart" min="1" max="100" value="1">
+    </fieldset>
+    <p>
+      List the items in each row, one line per row.  Each row consists of a set
+      of glosses separated by a space.
+    </p>
+      <label for="glossInput">Glosses:</label><br>
```
Comment (cindyli) [inline apps/palette-generator/palette-generator.html:54]: The label for the text area to receive BCI-AV-ID, glosses, SVGs etc should no longer be "Glosses" since it accepts a lot more than glosses.

File: apps/palette-generator/paletteJsonGenerator.ts [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+ *
+ * You may obtain a copy of the License at
+ * https://github.com/inclusive-design/adaptive-palette/blob/main/LICENSE
+ */
+import { v4 as uuidv4 } from "uuid";
+
+const BLANK_CELL_LABEL = "BLANK";
+const SVG_PREFIX = "SVG:";
+const SVG_SUFFIX = ":SVG";
+const LABEL_MARKER = "LABEL:";
```
Comment (cindyli) [inline apps/palette-generator/paletteJsonGenerator.ts:16]: When I input this line in the "Glosses" text area:

```
BLANK 18267LABEL:aaa induction
```
to alter the label for a symbol. I got this error in the error section: "BciAvId not found for label: induction". Moving the "LABEL:" marker to the last "induction" gloss reports the same error.

**Review by cindyli — CHANGES_REQUESTED** [review]

File: README.md [hunk 3]
```diff
@@ -132,3 +136,8 @@ command" and "Build output directory" settings:
 
 - Build command: `npm run build:client`
 - Build output directory:: `/dist/client`
+
+## Utility Scripts
+
+- [Load a document into a vector databbase (`scripts/loadDocIntoVectorDb.js`)](./scripts/loadDocIntoVectorDb.js)
+- [Generate a palette JSON file (`scripts/paletteJsonGenerator.js`)](./scripts/loadDocIntoVectorDb.js)
```
[the remark this answers is not in the document]
Comment (klown) [reply README.md:143]: Yup.  I'll remove it.

File: apps/palette-generator/README.md [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+project that run on localhost.  It
+is made available using the following command executed in the home directory:
+
+```text
+npm run serveAppsDemos
+```
+
+Once the development server is running, open this `localhost` url from within a
+browser:
+[`http://localhost:5173/apps/palette-generator/palette-generator.html`](http://localhost:5173/apps/palette-generator/palette-generator.html)
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/README.md:15]: Good idea.

File: apps/palette-generator/README.md [hunk 5]
```diff
@@ -0,0 +1,141 @@
...
+the top left corner of the palette.  The default values are row 1, column 1, and
+they refer to the position of the first cell in the first row of the palette.
+
+### Glosses
+
+The next section, "Glosses" is a text area where each line of text describes one
+row of the palette cells.  Spaces are used for separate cells so the text used
+to specify a cell cannot contain any spaces.
+
+There are three types of input allowed here:
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/README.md:35]: Yes.

File: apps/palette-generator/README.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
...
+
+For the case of a single word or single BCI AV ID, the label to display can be
+given by immediately appending "LABEL:*label\_text*" to the word or number. The
+*label\_text* will be used as the label for the cell instead of the usual gloss
+for the symbol.  Any underscore characters in the *label\_text* will be replaced
+with a single space.   For example "clothLABEL:cross_hatches" will result in the
+the label "cross hatches"; and "23641LABEL:ruin" will give "ruin" as the label
+instead of the official gloss "ruin,wreck,wreckage\_(building)\_(1)".
+
+The SVG string is marked by "SVG:" at the beginning of the string and ":SVG" at
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/README.md:52]: I've added two examples. But, I am not going to explain the full syntax of svg builder strings.  That should be in the svg-builder project, but there is no documentation there.  And, I don't know the full specification.  :-(

File: apps/palette-generator/README.md [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+
+For each match found, a cell is constructed and added to the appropriate row and
+the entire palette is shown on screen.  It can be inspected for errors.
+Corrections can be made in the glosses area and a new palette generated by
+clicking the "Generate palette" button again.
+
+### Save Palette
+
+Clicking the "Save palette" button will export a palette definition JSON file
+and save it into the user's "Downloads" folder.  The name of the file will be
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/README.md:88]: Good question. The code itself does not include the name "Downloads" or the name of any other folder.  An url is created that the browser understands, but is meaningless to me. The file ends up in the folder that has been set as the browser's download folder.  Users can set this to something other than "Downloads", in which case the palette will be saved based on that browser setting.

I will reword it to say, '.. and saved to disk where the browser saves file downloads, typically the "Downloads" folder.'

File: apps/palette-generator/palette-generator.ts [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+import { html } from "htm/preact";
+import { initAdaptivePaletteGlobals, adaptivePaletteGlobals } from "../../src/client/GlobalData";
+import { Palette } from "../../src/client/Palette";
+import { processPaletteLabels, fetchBlissGlossJson } from "./paletteJsonGenerator";
+import "../../src/client/index.scss";
+
+// Initialize any globals used elsewhere in the code.
+await initAdaptivePaletteGlobals("paletteDisplay");
+await fetchBlissGlossJson();
+let currentPaletteName = "";
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/palette-generator.ts:22]: I found a bug here and fixed it, but I have not addressed your comment fully:

> The save button doesn't save the given palette name.

It does after the fix, but there is still an edge case where (1) the palette is generated, (2) the user changes the name in the text field, and then (3) clicks the "Save" button.  In that case, the file is saved according the name in the text field when the palette was generated, not the just-changed text field  For this case, there needs to be an event handler for watching changes in the text field and adjusting various objects.  I'm working on that.

> The clear button doesn't clear the palette name too.

I don't think it should clear the name in the text field.  If a palette has been generated, but there are mistakes, and all the user wants to do is clear it, fix the mistake(s), and re-generate, then they should not have to re-enter the name. Their intent is to change the contents of the palette, not the name.  When they do want to change the name, then it's relatively simple to click on the text field and give it a new name.

File: apps/palette-generator/paletteJsonGenerator.ts [hunk 9]
```diff
@@ -0,0 +1,252 @@
...
+
+// Configurable options -- ideally provided by the UI
+const palette_name = "No name Palette";
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/paletteJsonGenerator.ts:19]: Good idea.  I will get the contents of the name text field and pass it via the generator function.

File: apps/palette-generator/paletteJsonGenerator.ts [hunk 10]
```diff
@@ -0,0 +1,252 @@
...
+const type = "ActionBmwCodeCell";
+
+// Special encodings for labels that don't follow the standard mapping
+const specialEncodings = {
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/paletteJsonGenerator.ts:23]: I'm inclined to remove it now.  But, just in case, what did you intend it for? Do the SVG: and LABEL: features cover that purpose?

File: apps/palette-generator/palette-generator.html (hunk 1 above)
Comment (klown) [reply apps/palette-generator/palette-generator.html:54]: Right, I'll come up with a better name.

File: apps/palette-generator/paletteJsonGenerator.ts (hunk 2 above)
Comment (klown) [reply apps/palette-generator/paletteJsonGenerator.ts:16]: That string specifies three cells:  a blank one, one for a plus sign labelled "aaa" and a cell whose gloss is "induction" or contains the word induction.  "Induction" does not match any gloss and that is why there is an error.  Checking blissary.com with "induction" gives "No results found."  The result is as expected.  Moving the LABEL marker won't change the result.

There was some text in the README explaining that spaces are used as a separator to indicate the string to use for a cell.  I've added more warnings to that effect including that LABEL: strings cannot contain spaces and to use an underscore if a space is desired.  Specifically, if the label for the plus sign was supposed to be "aaa induction", then you write it this way:

`BLANK 18267LABEL:aaa_induction`

That gives two cells, a blank and a plus with the label "aaa induction".

File: apps/palette-generator/palette-generator.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (cindyli) [reply apps/palette-generator/palette-generator.ts:22]: > I don't think it should clear the name in the text field.

Thanks for the clarification, and I agree.

File: apps/palette-generator/paletteJsonGenerator.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (cindyli) [reply apps/palette-generator/paletteJsonGenerator.ts:23]: I agree this special handling can be removed.

File: apps/palette-generator/paletteJsonGenerator.ts (hunk 2 above)
Comment (cindyli) [reply apps/palette-generator/paletteJsonGenerator.ts:16]: You are right. The error is reported correctly. I mistakenly thought it was related to the "LABEL" marker.

File: apps/palette-generator/README.md [hunk 11]
```diff
@@ -0,0 +1,141 @@
...
+with a single space.   For example "clothLABEL:cross_hatches" will result in the
+the label "cross hatches"; and "23641LABEL:ruin" will give "ruin" as the label
+instead of the official gloss "ruin,wreck,wreckage\_(building)\_(1)".
+
+The SVG string is marked by "SVG:" at the beginning of the string and ":SVG" at
+the end. What lies between is a set of comma separated BCI AVI IDs, and quoted
+strings that are parsed by the svg builder.  If the string "LABEL:*label\_text"
+occurs immediately before the closing ":SVG" marker, it will be used as the
+label for the cell, and is truncated from the rest of the string before the
+remainder is sent to the builder.  The rules for the syntax of this "LABEL" is
```
Comment (cindyli) [inline apps/palette-generator/README.md:57]: I tried with using "LABEL" with "SVG" marker. This string `SVG:14183,"/",25777,"/","W8W:0,8":SVGLABEL:aaa` leads to an error: "BciAvId not found for label: SVG:14183,"/",25777,"/","W8W:0,8":SVG". Did I combine the "SVG" and "LABEL" correctly?

**Comment by cindyli** [comment]
It would be helpful for users if you could add a small helper section next to the "Glosses" text field on the app interface. This area could provide quick explanations or examples of available markers like SVG, LABEL, and BLANK. I found I often doubt if I'm using these markers correctly and have to check README frequently.

File: apps/palette-generator/paletteJsonGenerator.ts [hunk 12]
```diff
@@ -0,0 +1,252 @@
...
+  // "specialLabel2": [12345, "/", 23456]
+};
+
+// End of configurable options
+
+let bliss_gloss;
+export async function fetchBlissGlossJson () {
+  // Read and parse the Bliss gloss JSON file
+  try {
+    const fetchResponse = await fetch("https://raw.githubusercontent.com/cindyli/baby-bliss-bot/refs/heads/feat/bmw/data/bliss_symbol_explanations.json");
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/paletteJsonGenerator.ts:36]: Good idea. I found a way to merge that one json file across the two repositories so the history is maintained.

File: apps/palette-generator/README.md [hunk 13]
```diff
@@ -0,0 +1,141 @@
...
+item in the "Glosses" input text area.  Each item is listed separately where the
+head of the listing is a copy of the input word.  Under this word is a list of
+BCI AV IDs and their glosses where a word in the gloss matched.  For example, if
+the input word was "cloth", two gloss entries are reported.  One is a symbol for
+cloth in the sense of fabric with BCI AV ID 13365, and the other is a symbol for
+drying rack, BCI AV ID 26161:
+
+```text
+cloth
+    13365
```
[the remark this answers is not in the document]
Comment (klown) [reply apps/palette-generator/README.md:111]: I tried that out and it looks better -- thanks!

File: apps/palette-generator/README.md (hunk 11 above)
Comment (klown) [reply apps/palette-generator/README.md:57]: Quoting from the README with emphasis:
> If the string "LABEL:*label\_text" occurs _immediately before the closing ":SVG" marker_, 

So your example should be `SVG:14183,"/",25777,"/","W8W:0,8","LABEL:aaa":SVG`

However, that's different from the two other cases of an ID or a gloss-string, where the `LABEL:` is appended to the ID or string.  I think it should be the same regardless of the input.  I have re-written the processing so that your example works, and the old way is now illegal.

**Comment by klown** [comment]
@cindyli  wrote:

> It would be helpful for users if you could add a small helper section next to the "Glosses" text field on the app interface. This area could provide quick explanations or examples of available markers like SVG, LABEL, and BLANK. I found I often doubt if I'm using these markers correctly and have to check README frequently.

I added one but in a way that allows users to pop it open as needed and then hide it.  I used the `details/summary` technique.  I have written and re-written the text for it, and am still not satisfied.  But, I have to stop at some point, so here it is.

I think I addressed all of your comments -- ready for another review, thanks.

File: apps/palette-generator/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (cindyli) [reply apps/palette-generator/README.md:15]: Since the entry page has been renamed to "index.html", the link and text should be changed to remove "palette-generator.html".

File: apps/palette-generator/palette-generator.ts [hunk 14]
```diff
@@ -0,0 +1,223 @@
...
+
+    const matchesForString = aMatch[matchString];
+    for (let i = 0; i < MAX_MATCHES_OUTPUT && i < matchesForString.length; i++) {
+      const match = matchesForString[i];
+      const dd = document.createElement("dd");
+      dl.append(dd);
+      dd.innerText = `${match.bciAvId}: ${match.label}`;
+      //       dd = document.createElement("dd");
+      //       dl.append(dd);
+      //       dd.innerText = match.label;
```
Comment (cindyli) [inline apps/palette-generator/palette-generator.ts:180]: Remove the commented code?

File: apps/palette-generator/paletteJsonGenerator.ts (hunk 10 above)
[the remark this answers is not in the document]
Comment (cindyli) [reply apps/palette-generator/paletteJsonGenerator.ts:22]: Remove the definition and the use of `specialEncodings`?

File: apps/palette-generator/paletteJsonGenerator.ts [hunk 15]
```diff
@@ -0,0 +1,241 @@
...
+    if ((label === gloss.description) || wordMatch.test(gloss.description)) {
+      matches.push({ bciAvId: parseInt(gloss.id), label: gloss.description });
+      console.log(`\tFound match: ${gloss.description}, bci-av-id: ${gloss.id}`);
+    }
+  }
+  // If no BCI AV ID is found, throw an error
+  if (matches.length === 0) {
+    throw new Error(`BciAvId not found for label: ${label}`);
+  }
+  console.log("");
```
Comment (cindyli) [inline apps/palette-generator/paletteJsonGenerator.ts:99]: Shall this line be removed?

**Review by cindyli — CHANGES_REQUESTED** [review]

File: apps/palette-generator/palette-generator.ts (hunk 14 above)
Comment (klown) [reply apps/palette-generator/palette-generator.ts:180]: Yes, no longer needed.

File: apps/palette-generator/paletteJsonGenerator.ts (hunk 15 above)
Comment (klown) [reply apps/palette-generator/paletteJsonGenerator.ts:99]: It was meant to output a blank line to separate the matches logged on the console.  But in FF it insists on printing `<empty string>` which is not a blank line.  I switched it to `console.log("\n");` and that does print a blank line.  Then I took it out completely, and the logged matches are readable even without a separating blank line.  So, no it's been removed.

**Comment by klown** [comment]
@cindyli, I've addressed you last comments.  I also added a new `<select>` element to allow users to choose the cell type.   However one type, the `ContentBmwEncoding` cell type, does not render well.  I am investigating.

Nonetheless, ready for another review.

**Comment by cindyli** [comment]
The pull request looks great. I will wait for the fix of the rendering of the `ContentBmwEncoding` cell type.

**Comment by klown** [comment]
I did manage to render `ContentBmwEncoding` by setting  up the `changeEncodingContents` signal correctly.  However, since the display area is a palette and not an content area, each cell is treated as a content area and all of the bliss symbols are shown in every cell.  It looks bizarre.

Then I realized that the palette json file to be saved should ever have cells of type `ContentBmwEncoding`.  So, since this is about _palette_ generation and not _content_ generation, the select is set up so that ContentBmwEncoding` is unavailable.

Let me know what you think.

**Comment by cindyli** [comment]
> Then I realized that the palette json file to be saved should never have cells of type `ContentBmwEncoding`. So, since this is about _palette_ generation and not _content_ generation, the select is set up so that `ContentBmwEncoding` is unavailable.

I agree this is fine.

**Review by cindyli — APPROVED** [review]
